serf/serf_test.go

```diff
@@ -276,6 +278,156 @@ func TestSerf_eventsLeave(t *testing.T) {
 		[]EventType{EventMemberJoin, EventMemberLeave})
 }
 
+func TestSerf_eventsLeave_avoidInfiniteLeaveRebroadcast(t *testing.T) {
+	// This test is a variation of the normal leave test that is crafted
+	// specifically to handle a situation where two unique leave events for the
+	// same node reach two other nodes in the wrong order which causes them to
+	// infinitely rebroadcast the leave event without updating their own
+	// lamport clock for that node.
+	ip1, returnFn1 := testutil.TakeIP()
+	defer returnFn1()
+
+	ip2, returnFn2 := testutil.TakeIP()
+	defer returnFn2()
+
+	ip3, returnFn3 := testutil.TakeIP()
+	defer returnFn3()
+
+	ip4, returnFn4 := testutil.TakeIP()
+	defer returnFn4()
+
+	testConfigLocal := func(t *testing.T, ip net.IP) *Config {
+		conf := testConfig(t, ip)
+		// Make the reap interval longer in this test
+		// so that the leave does not also cause a reap
+		conf.ReapInterval = 30 * time.Second
+		return conf
+	}
+
+	// Create the s1 config with an event channel so we can listen
+	eventCh := make(chan Event, 4)
+	s1Config := testConfigLocal(t, ip1)
+	s1Config.EventCh = eventCh
+
+	s2Config := testConfigLocal(t, ip2)
+	s2Config.RejoinAfterLeave = true
+	s2Addr := s2Config.MemberlistConfig.BindAddr
+	s2Name := s2Config.NodeName
+
+	s3Config := testConfigLocal(t, ip3)
+	// Allow s3 to drop joins in the future.
+	var s3DropJoins uint32
+	s3Config.messageDropper = func(t messageType) bool {
+		switch t {
+		case messageJoinType, messagePushPullType:
+			return atomic.LoadUint32(&s3DropJoins) == 1
+		default:
+			return false
+		}
+	}
+
+	s4Config := testConfigLocal(t, ip4)
+	// Allow s4 to drop joins in the future.
+	var s4DropJoins uint32
```

Comment:
We could use the same uint32 and dropper func for both s3 and s4 since we switch them both together anyway?